src/bridge/cockpitpcpmetrics.c

```diff
@@ -0,0 +1,1017 @@
+            {
+              problem = NULL;
+              if (message)
+                send_array (self, message);
+            }
+          else
+            {
+              g_message ("%s: couldn't read from archive: %s", self->name, pmErrStr (self->context));
+              problem = "internal-error";
+            }
+          cockpit_channel_close (COCKPIT_CHANNEL (self), problem);
+          if (message)
+            json_array_unref (message);
+          return FALSE;
+        }
+
+      meta = build_meta_if_necessary (self, result);
+      if (meta)
+        {
+          send_array (self, message);
+          json_array_unref (message);
+          message = NULL;
+
+          send_object (self, meta);
+          json_object_unref (meta);
+        }
+
+      if (message == NULL)
+          message = json_array_new ();
+      json_array_add_array_element (message, build_samples (self, result));
+
+      if (self->last)
+        {
+          pmFreeResult (self->last);
+          self->last = result;
+        }
+    }
+
+  if (message)
+    {
+      send_array (self, message);
+      json_array_unref (message);
+    }
+
+  return TRUE;
+}
+
+static void
+perform_load (CockpitPcpMetrics *self,
+              gint64 timestamp)
+{
+  struct timeval stamp;
+  pmResult *result;
+  int rc;
+
+  if (timestamp == 0)
+    {
+      stamp.tv_sec = 0x7fffffff;
+      if (pmSetMode (PM_MODE_BACK, &stamp, 0) != 0)
+        g_return_if_reached ();
+      rc = pmFetchArchive (&result);
+      if (rc < 0)
+        {
+          g_message ("%s: couldn't read from archive: %s", self->name, pmErrStr (self->context));
+          cockpit_channel_close (COCKPIT_CHANNEL (self), "internal-error");
+          return;
+        }
+      memcpy (&stamp, &result->timestamp, sizeof (stamp));
+      pmFreeResult (result);
+    }
+  else
+    {
+      stamp.tv_sec = (timestamp / 1000);
+      stamp.tv_usec = (timestamp % 1000) * 1000;
+    }
+
+  if (pmSetMode (PM_MODE_INTERP | PM_XTB_SET(PM_TIME_MSEC), &stamp, self->interval) != 0)
+    g_return_if_reached ();
+
+  if (on_idle_batch (self))
+    self->idler = g_idle_add (on_idle_batch, self);
+}
+
+static gboolean units_equal (pmUnits *a,
+                             pmUnits *b)
+{
+  return (a->scaleCount == b->scaleCount &&
+          a->scaleTime == b->scaleTime &&
+          a->scaleSpace == b->scaleSpace &&
+          a->dimCount == b->dimCount &&
+          a->dimTime == b->dimTime &&
+          a->dimSpace == b->dimSpace);
+}
+
+static gboolean units_convertible (pmUnits *a,
+                                   pmUnits *b)
+{
+  pmAtomValue dummy;
+  dummy.d = 0;
+  return pmConvScale (PM_TYPE_DOUBLE, &dummy, a, &dummy, b) >= 0;
+}
+
+static void
+cockpit_pcp_metrics_prepare (CockpitChannel *channel)
+{
+  CockpitPcpMetrics *self = COCKPIT_PCP_METRICS (channel);
+  const gchar *problem = "protocol-error";
+  JsonObject *options;
+  const gchar *source;
+  const gchar **instances = NULL;
```

Comment:
Yep, fixed.

However, the out value needs to be freed with g_free and not g_strfreev.